### ISSUE TITLE
AzureStorageContext.Context is not IStorageContext but a pointer back to the original AzureStorageContext

### ISSUE THREAD

**Issue by grecoe** [Azure/azure-powershell#12453]
Trying to create a container in a storage account by first getting a storage context with this: 

https://docs.microsoft.com/en-us/powershell/module/az.storage/new-azstoragecontext?view=azps-4.4.0

I read the documentation and it says that the return value is this:

https://docs.microsoft.com/en-us/dotnet/api/microsoft.windowsazure.commands.storage.azurestoragecontext?view=azurerm-ps

There it says that the property Context should return me an IStorageContext....but it doesn't, it just gives me bac the original AzureStorageContext

CODE:

    $res = New-AzStorageContext -StorageAccountName $storage_account -StorageAccountKey $storage_key

    Write-Host($res.GetType())
    Write-Host($res.Context.GetType())

    Write-Host(($res -eq $res.Context))

OUTPUT:
Microsoft.WindowsAzure.Commands.Storage.AzureStorageContext
Microsoft.WindowsAzure.Commands.Storage.AzureStorageContext
True

So, when I try to use it on 

https://docs.microsoft.com/en-us/powershell/module/azure.storage/new-azurestoragecontainer?view=azurermps-6.13.0

I get the following error 

New-AzureStorageContainer : Cannot bind parameter 'Context'. Cannot convert the "Microsoft.WindowsAzure.Commands.Storage.AzureStorageContext" value of type 
"Microsoft.WindowsAzure.Commands.Storage.AzureStorageContext" to type "Microsoft.Azure.Commands.Common.Authentication.Abstractions.IStorageContext".
At C:\gitenergykm\energy-knowledgemining-master\energy-knowledgemining-master\service\scripts\searchindex.ps1:113 char:75
+ ... AzureStorageContainer -Name $container_name -Context $storage_context
+                                                          ~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [New-AzureStorageContainer], ParameterBindingException
    + FullyQualifiedErrorId : CannotConvertArgumentNoMessage,Microsoft.WindowsAzure.Commands.Storage.Blob.Cmdlet.NewAzureStorageContainerCommand

**Comment by grecoe** [comment]
BTW, just set up this new laptop today, so the install is fresh. 

Version    Name                                Repository           Description
-------    ----                                ----------           -----------
5.1.2      Azure                               https://www.power... Microsoft Azure PowerShell - Service Management
0.5.0      Azure.AnalysisServices              https://www.power... Microsoft Azure PowerShell - Analysis Services server management
4.2.1      Azure.Storage                       https://www.power... Microsoft Azure PowerShell - Storage service cmdlets. Manages blobs, queues, tables an...
5.7.0      AzureRM                             https://www.power... Azure Resource Manager Module
0.6.6      AzureRM.AnalysisServices            https://www.power... Microsoft Azure PowerShell - Analysis Services
5.1.2      AzureRM.ApiManagement               https://www.power... Microsoft Azure PowerShell - Api Management service cmdlets for Azure Resource Manager   
0.1.3      AzureRM.ApplicationInsights         https://www.power... Microsoft Azure PowerShell - Application Insights service management cmdlets for Azure...
4.3.2      AzureRM.Automation                  https://www.power... Microsoft Azure PowerShell - Automation service cmdlets for Azure Resource Manager       
4.0.4      AzureRM.Backup                      https://www.power... Microsoft Azure PowerShell - Azure Backup service cmdlets for Azure Resource Manager     
4.0.6      AzureRM.Batch                       https://www.power... Microsoft Azure PowerShell - Batch service cmdlets for Azure Resource Manager
0.14.1     AzureRM.Billing                     https://www.power... Microsoft Azure PowerShell - Billing service cmdlets for Azure Resource Manager
4.2.2      AzureRM.Cdn                         https://www.power... Microsoft Azure PowerShell - CDN service cmdlets for Azure Resource Manager
0.9.4      AzureRM.CognitiveServices           https://www.power... Microsoft Azure PowerShell - Cognitive Services management cmdlets for Azure Resource ...
4.6.0      AzureRM.Compute                     https://www.power... Microsoft Azure PowerShell - Compute service cmdlets for Azure Resource Manager.  Mana...
0.3.1      AzureRM.Consumption                 https://www.power... Microsoft Azure PowerShell - Consumption service cmdlets for Azure Resource Manager      
0.2.5      AzureRM.ContainerInstance           https://www.power... Microsoft Azure PowerShell - Container Instance cmdlets for Azure Resource Manager       
1.0.4      AzureRM.ContainerRegistry           https://www.power... Microsoft Azure PowerShell - ContainerRegistry service cmdlets for Azure Resource Manager
4.2.2      AzureRM.DataFactories               https://www.power... Microsoft Azure PowerShell - DataFactories service cmdlets for Azure Resource Manager    
0.5.3      AzureRM.DataFactoryV2               https://www.power... Microsoft Azure PowerShell - DataFactories service cmdlets for Azure Resource Manager    
4.2.3      AzureRM.DataLakeAnalytics           https://www.power... Microsoft Azure PowerShell - Data Lake Analytics
5.2.0      AzureRM.DataLakeStore               https://www.power... Microsoft Azure PowerShell - Data Lake Store
4.0.4      AzureRM.DevTestLabs                 https://www.power... Microsoft Azure PowerShell - DevTestLabs service cmdlets for Azure Resource Manager      
4.1.2      AzureRM.Dns                         https://www.power... Microsoft Azure PowerShell - Dns service cmdlets for Azure Resource Manager
0.3.2      AzureRM.EventGrid                   https://www.power... Microsoft Azure PowerShell - EventGrid service cmdlets for Azure Resource Manager        
0.6.3      AzureRM.EventHub                    https://www.power... Microsoft Azure PowerShell - EventHubs service cmdlets for Azure Resource Manager        
4.1.2      AzureRM.HDInsight                   https://www.power... Microsoft Azure PowerShell - HDInsight service cmdlets for Azure Resource Manager        
4.0.4      AzureRM.Insights                    https://www.power... Microsoft Azure PowerShell - Insights service cmdlets for Azure Resource Manager
3.1.2      AzureRM.IotHub                      https://www.power... Microsoft Azure PowerShell - IotHub service cmdlets for Azure Resource Manager
4.3.0      AzureRM.KeyVault                    https://www.power... Microsoft Azure PowerShell - KeyVault service cmdlets for Azure Resource Manager
4.0.3      AzureRM.LogicApp                    https://www.power... Microsoft Azure PowerShell - LogicApp cmdlets for Azure Resource Manager
0.17.2     AzureRM.MachineLearning             https://www.power... Microsoft Azure PowerShell - Machine Learning Web Services cmdlets for Azure Resource ...
0.4.2      AzureRM.MachineLearningCompute      https://www.power... Microsoft Azure PowerShell - Machine Learning Compute cmdlets for Azure Resource Manager 
0.2.1      AzureRM.MarketplaceOrdering         https://www.power... Microsoft Azure PowerShell - MarketplaceOrdering agreements service cmdlets for Azure ...
0.9.2      AzureRM.Media                       https://www.power... Microsoft Azure PowerShell - Media service cmdlets for Azure Resource Manager
5.4.2      AzureRM.Network                     https://www.power... Microsoft Azure PowerShell - Network service cmdlets for Azure Resource Manager
4.1.1      AzureRM.NotificationHubs            https://www.power... Microsoft Azure PowerShell - NotificationHubs
4.3.2      AzureRM.OperationalInsights         https://www.power... Microsoft Azure PowerShell - OperationalInsights service cmdlets for Azure Resource Ma...
4.1.4      AzureRM.PowerBIEmbedded             https://www.power... Microsoft Azure PowerShell - Power BI Embedded service management cmdlets for Azure Re...
4.6.0      AzureRM.Profile                     https://www.power... Microsoft Azure PowerShell - Profile credential management cmdlets for Azure Resource ...
4.1.2      AzureRM.RecoveryServices            https://www.power... Microsoft Azure PowerShell - RecoveryServices cmdlets for Azure Resource Manager
4.1.2      AzureRM.RecoveryServices.Backup     https://www.power... Microsoft Azure PowerShell - Azure Backup service cmdlets for Azure Resource Manager     
0.2.4      AzureRM.RecoveryServices.SiteRec... https://www.power... Microsoft Azure PowerShell - SiteRecovery service cmdlets for Azure Resource Manager     
4.1.2      AzureRM.RedisCache                  https://www.power... Microsoft Azure PowerShell - RedisCache service cmdlets for Azure Resource Manager       
0.3.3      AzureRM.Relay                       https://www.power... Microsoft Azure PowerShell - Relay service cmdlets for Azure Resource Manager
5.5.2      AzureRM.Resources                   https://www.power... Microsoft Azure PowerShell - Azure Resource Manager cmdlets.  Manages subscriptions, t...
0.16.3     AzureRM.Scheduler                   https://www.power... Microsoft Azure PowerShell - Scheduler cmdlets for Azure Resource Manager
4.1.2      AzureRM.ServerManagement            https://www.power... Microsoft Azure PowerShell - ServerManagement cmdlets for Azure Resource Manager
0.6.4      AzureRM.ServiceBus                  https://www.power... Microsoft Azure PowerShell - ServiceBus service cmdlets for Azure Resource Manager       
0.3.4      AzureRM.ServiceFabric               https://www.power... Microsoft Azure PowerShell - Service Fabric cmdlets for Azure Resource Manager
5.0.6      AzureRM.SiteRecovery                https://www.power... Microsoft Azure PowerShell - SiteRecovery service cmdlets for Azure Resource Manager     
4.4.0      AzureRM.Sql                         https://www.power... Microsoft Azure PowerShell - Sql service cmdlets for Azure Resource Manager
4.2.3      AzureRM.Storage                     https://www.power... Microsoft Azure PowerShell - Storage service management cmdlets for Azure Resource Man...
4.0.4      AzureRM.StreamAnalytics             https://www.power... Microsoft Azure PowerShell - StreamAnalytics service cmdlets for Azure Resource Manager  
4.0.1      AzureRM.Tags                        https://www.power... Microsoft Azure PowerShell - Tags service cmdlets for Azure Resource Manager
4.0.3      AzureRM.TrafficManager              https://www.power... Microsoft Azure PowerShell - TrafficManager service cmdlets for Azure Resource Manager   
4.0.2      AzureRM.UsageAggregates             https://www.power... Microsoft Azure PowerShell - UsageAggregates service cmdlets for Azure Resource Manager  
4.2.2      AzureRM.Websites                    https://www.power... Microsoft Azure PowerShell - Websites service cmdlets for Azure Resource Manager
4.4.0      Az                                  PSGallery            Microsoft Azure PowerShell - Cmdlets to manage resources in Azure. This module is comp...
1.9.1      Az.Accounts                         PSGallery            Microsoft Azure PowerShell - Accounts credential management cmdlets for Azure Resource...
1.1.1      Az.Advisor                          PSGallery            Microsoft Azure PowerShell - Azure Advisor Cmdlets for Advisor in Windows PowerShell a...
1.1.3      Az.Aks                              PSGallery            Microsoft Azure PowerShell - Azure managed Kubernetes cmdlets for Windows PowerShell a...
1.1.4      Az.AnalysisServices                 PSGallery            Microsoft Azure PowerShell - Analysis Services cmdlets for Windows PowerShell and Powe...
2.0.1      Az.ApiManagement                    PSGallery            Microsoft Azure PowerShell - Api Management service cmdlets for Azure Resource Manager...
1.1.0      Az.ApplicationInsights              PSGallery            Microsoft Azure PowerShell - Application Insights management cmdlets for Azure Resourc...
1.3.7      Az.Automation                       PSGallery            Microsoft Azure PowerShell - Automation service cmdlets for Azure Resource Manager in ...
3.1.0      Az.Batch                            PSGallery            Microsoft Azure PowerShell - Batch service cmdlets for Azure Resource Manager in Windo...
1.0.3      Az.Billing                          PSGallery            Microsoft Azure PowerShell - Billing service cmdlets for Azure Resource Manager in Win...
1.4.3      Az.Cdn                              PSGallery            Microsoft Azure PowerShell - CDN service cmdlets for Azure Resource Manager in Windows...
1.5.0      Az.CognitiveServices                PSGallery            Microsoft Azure PowerShell - Cognitive Services management cmdlets for Azure Resource ...
4.2.1      Az.Compute                          PSGallery            Microsoft Azure PowerShell - Compute service cmdlets for Azure Resource Manager in Win...
1.0.3      Az.ContainerInstance                PSGallery            Microsoft Azure PowerShell - Container Instance cmdlets for Azure Resource Manager in ...
1.1.1      Az.ContainerRegistry                PSGallery            Microsoft Azure PowerShell - Container Registry service cmdlets for Azure Resource Man...
1.1.0      Az.DataBoxEdge                      PSGallery            Microsoft Azure PowerShell - DataBoxEdge service cmdlets for Azure Resource Manager in...
1.9.0      Az.DataFactory                      PSGallery            Microsoft Azure PowerShell - Data Factory service cmdlets for Azure Resource Manager i...
1.0.2      Az.DataLakeAnalytics                PSGallery            Microsoft Azure PowerShell - Data Lake Analytics in Windows PowerShell and PowerShell ...
1.2.8      Az.DataLakeStore                    PSGallery            Microsoft Azure PowerShell - Azure Data Lake Store cmdlets in Windows PowerShell and P...
1.0.0      Az.DataShare                        PSGallery            Microsoft Azure PowerShell - DataShare service cmdlets for Azure Resource Manager in W...
1.1.0      Az.DeploymentManager                PSGallery            PowerShell .Net Core Microsoft Azure PowerShell - Deployment Manager cmdlets for Azure...
1.0.0      Az.DesktopVirtualization            PSGallery            Microsoft Azure PowerShell: DesktopVirtualization cmdlets
1.0.2      Az.DevTestLabs                      PSGallery            Microsoft Azure PowerShell - DevTest Labs service cmdlets for Azure Resource Manager i...
1.1.2      Az.Dns                              PSGallery            Microsoft Azure PowerShell - DNS service cmdlets for Azure Resource Manager in Windows...
1.3.0      Az.EventGrid                        PSGallery            Microsoft Azure PowerShell - Event Grid service cmdlets for Azure Resource Manager in ...
1.5.0      Az.EventHub                         PSGallery            Microsoft Azure PowerShell - Event Hubs service cmdlets for Azure Resource Manager in ...
1.6.0      Az.FrontDoor                        PSGallery            Microsoft Azure PowerShell - Front Door service cmdlets for Azure Resource Manager in ...
1.0.1      Az.Functions                        PSGallery            Microsoft Azure PowerShell - Azure Functions service cmdlets for Azure Resource Manage...
3.3.1      Az.HDInsight                        PSGallery            Microsoft Azure PowerShell - HDInsight service cmdlets for Azure Resource Manager in W...
1.1.0      Az.HealthcareApis                   PSGallery            Microsoft Azure PowerShell - HealthcareApis management cmdlets for Azure Resource Mana...
2.5.0      Az.IotHub                           PSGallery            Microsoft Azure PowerShell - IoT Hub service cmdlets for Azure Resource Manager in Win...
2.0.0      Az.KeyVault                         PSGallery            Microsoft Azure PowerShell - Key Vault service cmdlets for Azure Resource Manager in W...
1.3.2      Az.LogicApp                         PSGallery            Microsoft Azure PowerShell - Logic Apps cmdlets for Azure Resource Manager in Windows ...
1.1.3      Az.MachineLearning                  PSGallery            Microsoft Azure PowerShell - Machine Learning Web Services cmdlets for Azure Resource ...
1.0.0      Az.Maintenance                      PSGallery            Microsoft Azure PowerShell - Maintenance cmdlets for Azure Resource Manager in Windows...
1.0.2      Az.ManagedServices                  PSGallery            Microsoft Azure PowerShell - ManagedServices cmdlets for Azure Resource Manager
1.0.2      Az.MarketplaceOrdering              PSGallery            Microsoft Azure PowerShell - Marketplace Ordering agreements service cmdlets for Azure...
1.1.1      Az.Media                            PSGallery            Microsoft Azure PowerShell - Media service cmdlets for Azure Resource Manager in Windo...
2.0.2      Az.Monitor                          PSGallery            Microsoft Azure PowerShell - Monitor service cmdlets for Azure Resource Manager in Win...
3.2.0      Az.Network                          PSGallery            Microsoft Azure PowerShell - Networking service cmdlets for Azure Resource Manager in ...
1.1.1      Az.NotificationHubs                 PSGallery            Microsoft Azure PowerShell - Notification Hubs cmdlets for Azure Resource Manager in W...
2.2.0      Az.OperationalInsights              PSGallery            Microsoft Azure PowerShell - Operational Insights service cmdlets for Azure Resource M...
1.3.1      Az.PolicyInsights                   PSGallery            Microsoft Azure PowerShell - Azure Policy Insights cmdlets for Windows PowerShell and ...
1.1.2      Az.PowerBIEmbedded                  PSGallery            Microsoft Azure PowerShell - Power BI Embedded service management cmdlets for Azure Re...
1.0.3      Az.PrivateDns                       PSGallery            Microsoft Azure PowerShell - Private DNS service cmdlets for Azure Resource Manager in...
2.11.0     Az.RecoveryServices                 PSGallery            Microsoft Azure PowerShell - Recovery Services cmdlets for Azure Resource Manager in W...
1.2.1      Az.RedisCache                       PSGallery            Microsoft Azure PowerShell - Redis Cache service cmdlets for Azure Resource Manager in...
1.0.3      Az.Relay                            PSGallery            Microsoft Azure PowerShell - Relay service cmdlets for Azure Resource Manager in Windo...
2.3.0      Az.Resources                        PSGallery            Microsoft Azure PowerShell - Azure Resource Manager and Active Directory cmdlets in Wi...
1.4.1      Az.ServiceBus                       PSGallery            Microsoft Azure PowerShell - Service Bus service cmdlets for Azure Resource Manager in...
2.1.0      Az.ServiceFabric                    PSGallery            Microsoft Azure PowerShell - Service Fabric cmdlets for Azure Resource Manager in Wind...
1.1.1      Az.SignalR                          PSGallery            Microsoft Azure PowerShell - Azure SignalR service commands for Windows PowerShell and...
2.9.0      Az.Sql                              PSGallery            Microsoft Azure PowerShell - SQL service cmdlets for Azure Resource Manager in Windows...
1.1.0      Az.SqlVirtualMachine                PSGallery            Microsoft Azure PowerShell - SqlVirtualMachine service cmdlets for Azure Resource Mana...
2.3.0      Az.Storage                          PSGallery            Microsoft Azure PowerShell - Storage service data plane and management cmdlets for Azu...
1.3.0      Az.StorageSync                      PSGallery            Microsoft Azure PowerShell - Storage Sync cmdlets in Windows PowerShell and PowerShell...
1.0.1      Az.StreamAnalytics                  PSGallery            Microsoft Azure PowerShell - Stream Analytics service cmdlets for Azure Resource Manag...
1.0.0      Az.Support                          PSGallery            Microsoft Azure PowerShell - Azure Support cmdlets for Azure Resource Manager in Windo...
1.0.4      Az.TrafficManager                   PSGallery            Microsoft Azure PowerShell - Traffic Manager service cmdlets for Azure Resource Manage...
1.11.0     Az.Websites                         PSGallery            Microsoft Azure PowerShell - App Service (Web Apps) service cmdlets for Azure Resource...
1.4.7      PackageManagement                   PSGallery            PackageManagement (a.k.a. OneGet) is a new way to discover and install software packag...

**Comment by dingmeng-xue** [comment]
Storage team, please help to look into this issue.

**Comment by blueww** [comment]
@grecoe

It seems you have load both old (AzureRm.* and Azure.*) and  new (Az.*) Azure Powershell modules into the Powershell console, so it will cause mess.

For you above scenario, you create Storage context with the new cmdlet `New-AzStorageContext` (in new module Az.Storage), but input the context to old cmdlet `New-AzureStorageContainer` (in old module Azure.Storage), so it doesn't work.
Please use the new cmdlet name `New-AzStorageContainer`, and it should work. Let me know if you still see any issue.

Besides that, is there any specific reason to keep the old Azure Powershell modules on your machine? If not, would you like to remove the old PSH modules? And the environment will be much more clean to avoid failure.

**Comment by grecoe** [comment]
@blueww yes, you are correct there were both sets of modules installed. Brand new machine and all I did was install the AZ modules so maybe the old AzureRM come pre-installed? 

I did uninstall the AzureRm and change over to New-AzStorageContainer and that seemed to solve the issue. I guess it was confusion on my part. 

Was able to move forward so I guess I'll close this? Not sue what the protocol is.
